_format/1.1/index.md

```diff
@@ -1839,3 +1877,572 @@ An error object **MAY** have the following members:
+  expressed as an object that **MAY** contain any of the following members:
+  - `type`: resource `type`
+  - `id` or `lid`: resource identifier or local identifier
+  - `relationship`: the name of the relationship
+
+* `params`: all of the query parameters allowed for singular requests can be
+  represented in JSON form in an operation's `params` object. Possible members
+  include:
+  - `include`
+  - `sort`
+  - `filter`
+  - `page`
+  - `fields`
+
+* `links`: a [links object](#document-links) related to the operation's primary
+  data.
+
+* `meta`: a [meta object](#document-meta) that contains non-standard
+  meta-information about the operation.
+
+Different members are required for processing different types of operations, as
+described below.
+
+> Note: `op` codes are used instead of the HTTP methods prescribed for singular
+  requests. The values `"get"`, `"add"`, `"update"`, and `"remove"` align with
+  the HTTP methods `GET`, `POST`, `PATCH`, and `DELETE` used for singular
+  requests.
+
+### <a href="#processing-operations" id="processing-operations" class="headerlink"></a> Processing Operations
+
+A server **MUST** perform operations in the order they appear in the
+`operations` array.
+
+All operations in a request **MUST** be performed atomically, so that a failure
+to perform any operation **MUST** invalidate any effects of preceding
+operations.
+
+A successful operation's `data` and `included` members **MUST** conform with the
+requirements for the analogous singular request and response.
+
+A server **MUST** respond to successful operations requests with `200 OK` if
+any operations are required to have `data` in a response. An array of resultant
+operation objects **MUST** be returned that correspond positionally to the
+requested operations.
+
+A server **MAY** respond to successful requests that include operations with
+`204 No Content` and no response document if no operations are required to
+return `data`.
+
+If any operation in a request fails, the server **MUST** respond as described
+in [Errors].
+
+> Note: An empty object (`{}`) is a valid operation object, and thus can be
+returned as a result to an operation request.
+
+> Note: The server can decide which operations and target resources are allowed
+  at each endpoint. Servers may use resource-specific endpoints like `/articles`
+  and `/authors` or generic endpoints for all resources like `/operations` or
+  `/bulk`.
+
+### Fetching Data
+
+An operation for fetching a resource, resource collection, resource identity, or
+resource identity collection **MUST** include an `op` code of `"get"`.
+
+#### Fetching Resources
+
+An operation that fetches an individual resource **MUST** target that resource
+either through the request endpoint or the operation's `ref` object.
+
+For example:
+
+```json
+PATCH /bulk HTTP/1.1
+Host: example.org
+Content-Type: application/vnd.api+json
+
+{
+  "operations": [{
+    "op": "get",
```

Comment:
I'm wondering if it should be forbidden to do operations requests that only fetch resources, because this severely hurts cacheability. Arguably, that's a choice of the client. But if the client uses tooling that uses operations requests for everything, that means they _never_ are able to reuse cached responses!

IOW: should operations requests have at least one mutation? Allowing reading data _after_ a mutation makes sense: to update the client's data store to the post-mutation state.

Reply:
> IOW: should operations requests have at least one mutation?

I think we can make this a recommendation but not a formal requirement.

Reply:
I suspect that many client implementations will be tempted to use `operations` for _all_ requests in pursuit some perceived simplification (they will not need to manage specific resource URLs, for example).

However, `PATCH` requests are extremely difficult to cache by nature. API-providers may be reluctant to implement this version of the spec if they are concerned that they will no longer be able to cache most reads at the edge or if it might require re-tooling for things like rate-limiting an/or API gateway quotas.

A statement like:

> Client implementations **SHOULD NOT** use operations requests where atomicity of the operations is not required.
> > NOTE: Clients should prefer multiple singular requests sent asynchronously to operations requests when their operations do not need to succeed or fail as a single transaction.

Would send a clear signal that there will be both a performance enhancement for the client (in theory, unrelated parallel PATCHes should be faster than the same ones performed serially). It might also put API providers' minds at ease.

For reference, [RFC5789 Patch Method for HTTP](https://tools.ietf.org/html/rfc5789) specification:

>  A response to [PATCH] is only cacheable if it
>  contains explicit freshness information (such as an Expires header or
>  "Cache-Control: max-age" directive) as well as the Content-Location
>  header matching the Request-URI, indicating that the PATCH response
>  body is a resource representation.  A cached PATCH response can only
>  be used to respond to subsequent GET and HEAD requests; it MUST NOT
>  be used to respond to other methods (in particular, PATCH).

Reply:
@gabesullice I agree that we should make a pretty clear statement about caching + `PATCH` and recommend support for singular requests in addition to operations. Thanks for your suggestions!